Packet++/header/GvcpLayer.h

```diff
@@ -0,0 +1,772 @@
+		uint32_t subnetMask = 0;
+		char padding4[12] = { 0 };
+		uint32_t gateway = 0;
+	};
+	static_assert(sizeof(GvcpForceIpBody) == internal::kGvcpForceIpBodyLength,
+	              "GVCP force IP command body size should be 56 bytes");
+#pragma pack(pop)
+
+	/**
+	 * @class GvcpLayer
+	 * A class representing the GigE Vision protocol(GVCP).
+	 * The class is implemented according to the GigE Vision specification 2.0.
+	 * @see https://en.wikipedia.org/wiki/GigE_Vision
+	 * @note The class cannot be instantiated directly.
+	 */
+	class GvcpLayer : public Layer
+	{
+	public:
+		/**
+		 * A static method that checks whether the port is considered as GVCP
+		 * @param[in] port The port number to be checked
+		 */
+		static bool isGvcpPort(uint16_t port)
+		{
+			return port == 3956;
+		}
+
+		/**
+		 * @brief Get the magic number
+		 * @return uint8_t The magic number
+		 */
+		static bool verifyRequest(const uint8_t* data)
+		{
+			return data[0] == internal::kGvcpMagicNumber;
+		};
+
+		/**
+		 * @brief Construct a new GvcpLayer object
+		 * @param[in] data A pointer to the raw data
+		 * @param[in] dataLen Size of the data in bytes
+		 * @param[in] prevLayer A pointer to the previous layer
+		 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+		 * @return GvcpLayer* A pointer to the constructed GvcpLayer object
+		 */
+		static GvcpLayer* parseGvcpLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
+
+	protected:
+		GvcpLayer() = default;
+
+		/**
+		 * @brief Construct a new GvcpLayer object
+		 * @param[in] data A pointer to the raw data
+		 * @param[in] dataLen Size of the data in bytes
+		 * @param[in] prevLayer A pointer to the previous layer
+		 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+		 */
+		GvcpLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
+
+		// implement Layer's abstract methods
+		void parseNextLayer() override
+		{}
+
+		// implement Layer's abstract methods
+		void computeCalculateFields() override
+		{}
+
+		// implement Layer's abstract methods
+		OsiModelLayer getOsiModelLayer() const override
+		{
+			return OsiModelLayer::OsiModelApplicationLayer;
+		}
+	};
+
+	/// @brief Gvcp request layer
+	class GvcpRequestLayer : public GvcpLayer
+	{
+	public:
+		/**
+		 * @brief Construct a new GvcpLayer object
+		 * @param[in] data A pointer to the raw data
+		 * @param[in] dataLen Size of the data in bytes
+		 * @param[in] prevLayer A pointer to the previous layer
+		 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+		 */
+		GvcpRequestLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
+
+		/**
+		 * @brief Construct a new GvcpRequestLayer object
+		 * @param[in] data A pointer to the data including the header and the payload
+		 * @param[in] dataSize The size of the data in bytes
+		 */
+		GvcpRequestLayer(const uint8_t* data, size_t dataSize);
```

Comment:
Why do we need this c'tor if we have the other c'tor below?

Reply:
What if the user wants to build the layer from the bytes of the whole layer? The following constructor is for building the layer with the header information and the payload data.

Reply:
That's usually not required. If users have raw data they create a `pcpp::Packet` out of it and use this c'tor: `GvcpRequestLayer(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet)`